File: packages/sitemark_system_api/android/src/main/kotlin/io/github/wikg1018/sitemark/system/CaptureSessionPolicy.kt
package io.github.wikg1018.sitemark.system

import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri

/**
 * Durable camera-session helpers. Clearing pending capture state uses
 * [SharedPreferences.Editor.commit] so a process death cannot resurrect a
 * stale target after finish. URI grants are revoked with the same READ|WRITE
 * flags used when launching the system camera.
 */
internal object CaptureSessionPolicy {
    const val PREFERENCES = "sitemark_capture_recovery"
    const val KEY_CAPTURE_ID = "capture_id"
    const val KEY_CAPTURE_PATH = "capture_path"
    const val URI_GRANT_FLAGS =
        Intent.FLAG_GRANT_READ_URI_PERMISSION or Intent.FLAG_GRANT_WRITE_URI_PERMISSION

    fun captureContentUri(packageName: String, captureId: String): Uri =
        Uri.Builder()
            .scheme("content")
            .authority("$packageName.capture")
            .appendPath("capture")
            .appendPath(captureId)
            .build()

    fun shouldRevoke(captureId: String?): Boolean = !captureId.isNullOrEmpty()

    fun clearPending(editor: SharedPreferences.Editor): Boolean =
        editor.remove(KEY_CAPTURE_ID).remove(KEY_CAPTURE_PATH).commit()
}

File: packages/sitemark_system_api/ios/Classes/CaptureSessionPolicy.swift
import Foundation

/// Synchronous, durable key-value editing for pending camera-session
/// state. Android clears pending state via `SharedPreferences.commit` so a
/// process death cannot resurrect a stale target; iOS has no content-URI
/// grant to revoke and no SharedPreferences, so this protocol plus the key
/// constants are the whole portable contract. The 2b production
/// implementation backs it with `UserDefaults` + an immediate synchronize.
public protocol CaptureStateStore: AnyObject {
    /// Removes the given keys durably and synchronously. Returns false when
    /// the durable write failed (mirrors `SharedPreferences.commit`).
    func removeValues(forKey keys: [String]) -> Bool
}

public enum CaptureSessionPolicy {
    /// Suite name for the pending-capture state, mirroring the Android
    /// preference file name so diagnostics stay comparable across platforms.
    public static let preferencesName = "sitemark_capture_recovery"
    public static let keyCaptureId = "capture_id"
    public static let keyCapturePath = "capture_path"

    /// Clears the pending capture state synchronously and durably.
    public static func clearPending(_ store: CaptureStateStore) -> Bool {
        store.removeValues(forKey: [keyCaptureId, keyCapturePath])
    }
}
